### PR TITLE
feat: add ability to include dimensions per-metric

### PR DESCRIPTION
**Pull request by austince**
Been using the exclude feature that we put in yesterday (#28) and liking it, but found it would be useful to have a whitelist option as well. Tried to match inclusion/exclusion logic of `INCLUDE_METRICS` / `EXCLUDE_METRICS`.

### CODE REVIEW

File: README.yaml [hunk 1]
```diff
@@ -62,7 +62,8 @@ usage: |-
...
   | include_metrics                | INCLUDE_METRICS                | Only publish the specified metrics (comma-separated list of glob patterns)                                                                                               |
   | exclude_metrics                | EXCLUDE_METRICS                | Never publish the specified metrics (comma-separated list of glob patterns)                                                                                              |
-  | exclude_dimensions_for_metrics | EXCLUDE_DIMENSIONS_FOR_METRICS | Dimensions to exclude for metrics (semi-colon-separated key values of comma-separated dimensions of METRIC=dim1,dim2;, e.g. 'flink_jobmanager=job,host;zk_up=host,pod;') |
+  | include_dimensions_for_metrics | INCLUDE_DIMENSIONS_FOR_METRICS | Dimensions to include for metrics (semi-colon-separated key values of comma-separated dimensions of METRIC=dim1,dim2;, e.g. 'flink_jobmanager=job_id')                   |
```
Comment (aknysh) [inline README.yaml]:
```suggestion
  | include_dimensions_for_metrics | INCLUDE_DIMENSIONS_FOR_METRICS | Dimensions to include for metrics (semi-colon-separated key values of comma-separated dimensions of METRIC=dim1,dim2;, e.g. 'flink_jobmanager=job_id'). Only publish the specified dimensions                   |
```

File: README.yaml [hunk 2]
```diff
@@ -62,7 +62,8 @@ usage: |-
...
+  | exclude_dimensions_for_metrics | EXCLUDE_DIMENSIONS_FOR_METRICS | Dimensions to exclude for metrics (semi-colon-separated key values of comma-separated dimensions of METRIC=dim1,dim2;, e.g. 'flink_jobmanager=job,host;zk_up=host,pod')  |
```
Comment (aknysh) [inline README.yaml]:
```suggestion
  | exclude_dimensions_for_metrics | EXCLUDE_DIMENSIONS_FOR_METRICS | Dimensions to exclude for metrics (semi-colon-separated key values of comma-separated dimensions of METRIC=dim1,dim2;, e.g. 'flink_jobmanager=job,host;zk_up=host,pod')  Never publish the specified dimensions |
```

File: main.go [hunk 3]
```diff
@@ -29,6 +30,7 @@ var (
 	replaceDimensions           = flag.String("replace_dimensions", os.Getenv("REPLACE_DIMENSIONS"), "replace dimensions specified by NAME=VALUE,...")
 	includeMetrics              = flag.String("include_metrics", os.Getenv("INCLUDE_METRICS"), "Only publish the specified metrics (comma-separated list of glob patterns, e.g. 'up,http_*')")
 	excludeMetrics              = flag.String("exclude_metrics", os.Getenv("EXCLUDE_METRICS"), "Never publish the specified metrics (comma-separated list of glob patterns, e.g. 'tomcat_*')")
+	includeDimensionsForMetrics = flag.String("include_dimensions_for_metrics", os.Getenv("INCLUDE_DIMENSIONS_FOR_METRICS"), "Dimensions to include for metrics (semi-colon-separated key values of comma-separated dimensions of METRIC=dim1,dim2;, e.g. 'flink_jobmanager=job_id')")
```
Comment (aknysh) [inline main.go]:
```suggestion
	includeDimensionsForMetrics = flag.String("include_dimensions_for_metrics", os.Getenv("INCLUDE_DIMENSIONS_FOR_METRICS"), "Dimensions to include for metrics (semi-colon-separated key values of comma-separated dimensions of METRIC=dim1,dim2;, e.g. 'flink_jobmanager=job_id'). Only publish the specified dimensions")
```

File: main.go [hunk 4]
```diff
@@ -29,6 +30,7 @@ var (
...
 	excludeDimensionsForMetrics = flag.String("exclude_dimensions_for_metrics", os.Getenv("EXCLUDE_DIMENSIONS_FOR_METRICS"), "Dimensions to exclude for metrics (semi-colon-separated key values of comma-separated dimensions of METRIC=dim1,dim2;, e.g. 'flink_jobmanager=job,host;zk_up=host,pod;')")
```
Comment (aknysh) [inline main.go]:
```suggestion
	excludeDimensionsForMetrics = flag.String("exclude_dimensions_for_metrics", os.Getenv("EXCLUDE_DIMENSIONS_FOR_METRICS"), "Dimensions to exclude for metrics (semi-colon-separated key values of comma-separated dimensions of METRIC=dim1,dim2;, e.g. 'flink_jobmanager=job,host;zk_up=host,pod;'). Never publish the specified dimensions")
```

**Comment by austince** [comment]
Thanks for the quick review @aknysh. Sounds good, I'll make sure of that.

**Comment by austince** [comment]
```go
type StringSet map[string]bool

func set(slice []string) StringSet {
	boolMap := make(StringSet, len(slice))

	for i := range slice {
		boolMap[slice[i]] = true
	}

	return boolMap
}

func shouldIncludeDimension(dimName string, includeSet, excludeSet StringSet) bool {
	// blacklist first
	if excludeSet != nil && excludeSet[dimName] {
		return false
	}
	// if no whitelist, keep it as it passed the blacklist
	if includeSet == nil {
		return true
	}
	// otherwise, check the whitelist
	return includeSet[dimName]
}

func main() {
	println(shouldIncludeDimension("dim1", nil, set([]string{"dim1"})))                   // false, in exclude set
	println(shouldIncludeDimension("dim2", set([]string{"dim1"}), nil))                   // false, not in include set
	println(shouldIncludeDimension("dim3", set([]string{"dim1"}), set([]string{"dim2"}))) // false, not in include set yet also not in exclude set
	println(shouldIncludeDimension("dim1", nil, set([]string{"dim2"})))                   // true, no include set specified and not in exclude set
	println(shouldIncludeDimension("dim1", set([]string{"dim1"}), set([]string{"dim2"}))) // true, in the include set, not in exclude set
}
```

I think we should be good - here's a runnable example of the current logic.

**Comment by austince** [comment]
Thanks @aknysh - I'm not sure if I'll have time soon but might in a few weeks.

**Pull request merged**